src/dynamic_fmt.c

```diff
@@ -404,6 +408,46 @@ char *RemoveHEX(char *output, char *input)
 	return output;
 }
 
+// Add $HEX$ encoding for salts where needed.  For canonicalized pot storage, we encode anything
+// containing unprintables or field separators, but never anything else
+static char *AddHEX(char *ciphertext) {
+	static char out[1024];
+	char *cp;
+
+	if (strncmp(ciphertext, "$dynamic_", 9) && strncmp(ciphertext, "@dynamic=", 9))
+		return ciphertext;  // not a dynamic format, so we can not 'fix' it.
+
+	// ok, see if this is salted:
+	if (!(cp = strchr(&ciphertext[1], ciphertext[0])) || !(cp = strchr(&cp[1], '$')))
+		return ciphertext;  // not a salted format.
+
+	if (!strncmp(cp, "$HEX$", 5))
+		return ciphertext;  // already HEX$
+
+	char *s = cp++;
+	int need_hex = 0;
+	if (cp[strlen(cp) - 1] == ' ' || cp[strlen(cp) - 1] == '\t' )
+		need_hex = 1;
+	else
+	while (*++s) {
+		if (*s < ' ' || *s > 0x7e || *s == ':' || *s == '$') {
+			need_hex = 1;
+			break;
+		}
+	}
+
+	if (need_hex) {
+		char *cpx;
+		cpx = out;
+		cpx += sprintf(out, "%*.*sHEX$", (int)(cp-ciphertext), (int)(cp-ciphertext), ciphertext);
+		while (*cp)
+			cpx += sprintf(cpx, "%02x", (unsigned char)(*cp++));
```

Comment:
This makes the size of output larger than input's. Is `out` really large enough?

Reply:
What I do (think I) know is that many tests I've in other parts of dynamics assume (unhexed) stuff is shorter total than 512, correct or not... But again, I should probably revert this to the `mem_alloc_tiny`.

Reply:
I opted for realloc instead, in steps of 1K. In case we load gazillions of some dynamic hashes it could be a problematic waste to mem_alloc_tiny each of them.